### PR TITLE
*: fix some problems related to notNullFlag 

### LINKED ISSUE #18488: NOT_NULL Flag incorrectly set
### 1. Minimal reproduce step (Required)

````
CREATE TABLE `test` (
            `iD` bigint(20) NOT NULL,
            `INT_TEST` int(11) DEFAULT NULL
          ) ENGINE=InnoDB DEFAULT CHARSET=utf8 ROW_FORMAT=DYNAMIC;
INSERT INTO test VALUES (2, 10), (3, NULL);

 SELECT DISTINCT count(*), id + int_test as res FROM test  GROUP BY res ORDER BY res;
````

### 2. What did you expect to see? (Required)

````
mysql>  SELECT DISTINCT count(*), id + int_test as res FROM test  GROUP BY res ORDER BY res;
+----------+-----+
| count(*) | res |
+----------+-----+
|        1 | NULL|
|        1 |  12 |
+----------+-----+
2 rows in set (0.01 sec)
````
### 3. What did you see instead (Required)

````
mysql>  SELECT DISTINCT count(*), id + int_test as res FROM test  GROUP BY res ORDER BY res;
+----------+-----+
| count(*) | res |
+----------+-----+
|        1 | NULL |
|        1 |  12 |
+----------+-----+
2 rows in set (0.01 sec)
````

### 4. Affected version (Required)

````
commit a9177fe846bf2638261f64514ab4ebd0155cc60c (HEAD -> master, origin/master, origin/HEAD)
Author: Boqin Qin <bobbqqin@bupt.edu.cn>
Date:   Mon Jul 6 19:28:12 2020 +0800
````
### 5. Root Cause Analysis

An extra space is printed.

### LINKED ISSUE #19025: JSON type has an unexpected NOT_NULL flag.
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```mysql
create table t ( c_bit bit(10), c_int_d int, c_uint_d int unsigned, c_bigint_d bigint, c_ubigint_d bigint unsigned, c_float_d float, c_ufloat_d float unsigned, c_double_d double, c_udouble_d double unsigned, c_decimal decimal(6, 3), c_udecimal decimal(10, 3) unsigned, c_decimal_d decimal, c_udecimal_d decimal unsigned, c_datetime datetime(2), c_datetime_d datetime, c_time time(3), c_time_d time, c_date date, c_timestamp timestamp(4) DEFAULT CURRENT_TIMESTAMP(4), c_timestamp_d timestamp DEFAULT CURRENT_TIMESTAMP, c_char char(20), c_bchar char(20) binary, c_varchar varchar(20), c_bvarchar varchar(20) binary, c_text_d text, c_btext_d text binary, c_binary binary(20), c_varbinary varbinary(20), c_blob_d blob, c_set set('a', 'b', 'c'), c_enum enum('a', 'b', 'c'), c_json JSON, c_year year );
```

### 2. What did you expect to see? (Required)

```mysql
mysql> select c_json from t;
Field   1:  `c_json`
Catalog:    `def`
Database:   `test`
Table:      `t`
Org_table:  `t`
Type:       JSON
Collation:  binary (63)
Length:     4294967295
Max_length: 0
Decimals:   0
Flags:      BLOB BINARY 


0 rows in set (0.02 sec)
```

### 3. What did you see instead (Required)

```mysql
mysql> select c_json from t;
Field   1:  `c_json`
Catalog:    `def`
Database:   `test`
Table:      `t`
Org_table:  `t`
Type:       JSON
Collation:  binary (63)
Length:     4294967295
Max_length: 0
Decimals:   0
Flags:      NOT_NULL BINARY 


0 rows in set (0.01 sec)
```

### PR DESCRIPTION
**Pull request by wjhuang2016**
### What problem does this PR solve?

Issue Number: close #19025, close #18488 

Problem Summary:
This PR is based on https://github.com/pingcap/tidb/pull/19029. It fixes some problems related to the notNullFlag.

### What is changed and how it works?

Remove adding NotNullFlag in dumpFlag.
Add NotNullFlag for Null constant in DefaultTypeForValue.

### Check List 

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note 

```release-note
fix some problems related to notNullFlag 
```

### CODE REVIEW

**Review by lzmhhh123 — APPROVED** [review]

File: server/column.go [hunk 1]
```diff
@@ -76,9 +76,6 @@ func dumpFlag(tp byte, flag uint16) uint16 {
 	case mysql.TypeEnum:
 		return flag | uint16(mysql.EnumFlag)
 	default:
-		if mysql.HasBinaryFlag(uint(flag)) {
```
Comment (XuHuaiyu) [inline server/column.go:79]: Why remove this?

File: expression/constant.go [hunk 2]
```diff
@@ -71,6 +71,19 @@ type Constant struct {
 	collationInfo
 }
 
+// Clone implements Expression interface.
+func (c *Constant) Clone() Expression {
```
Comment (XuHuaiyu) [inline expression/constant.go:75]: Why change this?

File: expression/simple_rewriter.go [hunk 3]
```diff
@@ -58,7 +58,7 @@ func ParseSimpleExprCastWithTableInfo(ctx sessionctx.Context, exprStr string, ta
 	if err != nil {
 		return nil, err
 	}
-	e = BuildCastFunction(ctx, e, targetFt)
+	e = BuildCastFunction(ctx, e, targetFt.Clone())
```
Comment (XuHuaiyu) [inline expression/simple_rewriter.go:61]: Why change this?

File: server/column.go (hunk 1 above)
Comment (wjhuang2016) [reply server/column.go:79]: Because it's unreasonable.

File: expression/constant.go (hunk 2 above)
Comment (wjhuang2016) [reply expression/constant.go:75]: From  #19029, I think it's the right way for Clone. We should clone the retType too. Though it's not necessary for this PR.

File: expression/simple_rewriter.go (hunk 3 above)
Comment (wjhuang2016) [reply expression/simple_rewriter.go:61]: From #19029, I think it's reasonable not to let column and scalarFunction share the same ft. Though it's not necessary for this PR.

File: expression/constant.go (hunk 2 above)
Comment (XuHuaiyu) [reply expression/constant.go:75]: If this modification is not necessary for this PR, I think we can remove it.

**Review by XuHuaiyu — APPROVED** [review]
LGTM

**Pull request merged**